File: src/main/java/com/example/productapp/controller/ProductController.java
package com.example.productapp.controller;

import com.example.productapp.model.Product;
import com.example.productapp.model.ProductVariant;
import com.example.productapp.service.ProductFetchService;
import com.example.productapp.service.ProductService;
import org.springframework.stereotype.Controller;
import org.springframework.ui.Model;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestParam;

import java.math.BigDecimal;
import java.util.ArrayList;
import java.util.List;

@Controller
public class ProductController {

    private final ProductService productService;
    private final ProductFetchService productFetchService;

    public ProductController(ProductService productService, ProductFetchService productFetchService) {
        this.productService = productService;
        this.productFetchService = productFetchService;
    }

    @GetMapping("/")
    public String index(Model model) {
        model.addAttribute("productCount", productService.getProductCount());
        return "index";
    }

    @GetMapping("/products")
    public String loadProducts(Model model) {
        List<Product> products = productService.getAllProducts();
        model.addAttribute("products", products);
        return "fragments/product-table :: productTable";
    }

    @PostMapping("/products")
    public String addProduct(@RequestParam String title,
            @RequestParam(required = false) String vendor,
            @RequestParam(required = false) String productType,
            @RequestParam(required = false) BigDecimal price,
            @RequestParam(required = false) List<String> variantTitles,
            @RequestParam(required = false) List<String> variantSkus,
            @RequestParam(required = false) List<BigDecimal> variantPrices,
            @RequestParam(required = false) List<Boolean> variantAvailables,
            Model model) {

        List<ProductVariant> variants = new ArrayList<>();
        if (variantTitles != null) {
            for (int i = 0; i < variantTitles.size(); i++) {
                String vTitle = variantTitles.get(i);
                if (vTitle != null && !vTitle.isBlank()) {
                    ProductVariant v = new ProductVariant();
                    v.setTitle(vTitle);
                    v.setSku(variantSkus != null && i < variantSkus.size() ? variantSkus.get(i) : null);
                    v.setPrice(variantPrices != null && i < variantPrices.size() ? variantPrices.get(i) : price);
                    v.setAvailable(variantAvailables != null && i < variantAvailables.size() ? variantAvailables.get(i)
                            : true);
                    variants.add(v);
                }
            }
        }

        productService.addProduct(title, vendor, productType, price, variants);
        List<Product> products = productService.getAllProducts();
        model.addAttribute("products", products);
        return "fragments/product-table :: productTable";
    }

    @PostMapping("/products/fetch")
    public String fetchProducts(Model model) {
        productFetchService.fetchProductsManually();
        List<Product> products = productService.getAllProducts();
        model.addAttribute("products", products);
        return "fragments/product-table :: productTable";
    }
}

File: src/main/kotlin/com/example/productapp/controller/ProductController.kt
package com.example.productapp.controller

import com.example.productapp.model.ProductVariant
import com.example.productapp.service.ProductService
import org.springframework.http.HttpStatus
import org.springframework.stereotype.Controller
import org.springframework.ui.Model
import org.springframework.web.bind.annotation.*
import org.springframework.web.server.ResponseStatusException
import org.springframework.web.servlet.mvc.support.RedirectAttributes
import java.math.BigDecimal

@Controller
class ProductController(private val productService: ProductService) {

    @GetMapping("/")
    fun index(model: Model): String {
        model.addAttribute("productCount", productService.getProductCount())
        return "index"
    }

    @GetMapping("/products")
    fun loadProducts(model: Model): String {
        val products = productService.getAllProducts()
        model.addAttribute("products", products)
        return "fragments/product-table :: productTable"
    }

    @PostMapping("/products")
    fun addProduct(
        @RequestParam title: String,
        @RequestParam(required = false) vendor: String?,
        @RequestParam(required = false) productType: String?,
        @RequestParam(required = false) price: BigDecimal?,
        @RequestParam(required = false) variantTitles: List<String>?,
        @RequestParam(required = false) variantSkus: List<String>?,
        @RequestParam(required = false) variantPrices: List<BigDecimal>?,
        @RequestParam(required = false) variantAvailables: List<Boolean>?,
        model: Model
    ): String {
        val variants = mutableListOf<ProductVariant>()
        variantTitles?.forEachIndexed { i, vTitle ->
            if (vTitle.isNotBlank()) {
                val variant = ProductVariant(
                    title = vTitle,
                    sku = variantSkus?.getOrNull(i),
                    price = variantPrices?.getOrNull(i) ?: price,
                    available = variantAvailables?.getOrNull(i) ?: true
                )
                variants.add(variant)
            }
        }

        productService.addProduct(title, vendor, productType, price, variants)
        val products = productService.getAllProducts()
        model.addAttribute("products", products)
        return "fragments/product-table :: productTable"
    }

    // ===== Search =====

    @GetMapping("/search")
    fun searchPage(): String = "search"

    @GetMapping("/search/results")
    fun searchResults(@RequestParam(name = "q", defaultValue = "") query: String, model: Model): String {
        val products = productService.searchProducts(query)
        model.addAttribute("products", products)
        return "fragments/product-table :: productTable"
    }

    // ===== Edit Product =====

    @GetMapping("/products/{id}/edit")
    fun editProductPage(@PathVariable id: Long, model: Model): String {
        val product = productService.getProductById(id)
            .orElseThrow { ResponseStatusException(HttpStatus.NOT_FOUND, "Product not found") }
        model.addAttribute("product", product)
        return "edit-product"
    }

    @PostMapping("/products/{id}")
    fun updateProduct(
        @PathVariable id: Long,
        @RequestParam title: String,
        @RequestParam(required = false) vendor: String?,
        @RequestParam(required = false) productType: String?,
        @RequestParam(required = false) price: BigDecimal?,
        redirectAttributes: RedirectAttributes
    ): String {
        productService.updateProduct(id, title, vendor, productType, price)
        redirectAttributes.addFlashAttribute("successMessage", "Product updated successfully!")
        return "redirect:/products/$id/edit"
    }

    // ===== Delete Product =====

    @DeleteMapping("/products/{id}")
    fun deleteProduct(@PathVariable id: Long, model: Model): String {
        productService.deleteProduct(id)
        val products = productService.getAllProducts()
        model.addAttribute("products", products)
        return "fragments/product-table :: productTable"
    }

    // ===== Toggle Variant Availability =====

    @PutMapping("/variants/{id}/toggle-availability")
    fun toggleVariantAvailability(@PathVariable id: Long, model: Model): String {
        val variant = productService.toggleVariantAvailability(id)
            .orElseThrow { ResponseStatusException(HttpStatus.NOT_FOUND, "Variant not found") }
        model.addAttribute("variant", variant)
        return "fragments/variant-availability :: variantAvailability"
    }
}
